drivers/i2s/i2s_esp32.c

```diff
@@ -0,0 +1,948 @@
+	clock_control_subsys_t clock_subsys;
+};
+
+struct i2s_esp32_data {
+	i2s_hal_clock_info_t clk_info;
+	struct i2s_esp32_stream rx;
+	struct i2s_esp32_stream tx;
+};
+
+uint32_t i2s_esp32_get_source_clk_freq(i2s_clock_src_t clk_src)
+{
+	uint32_t clk_freq = 0;
+
+	esp_clk_tree_src_get_freq_hz(clk_src, ESP_CLK_TREE_SRC_FREQ_PRECISION_CACHED, &clk_freq);
+	return clk_freq;
+}
+
+static esp_err_t i2s_esp32_calculate_clock(const struct i2s_config *i2s_cfg, uint8_t channel_length,
+					   i2s_hal_clock_info_t *i2s_hal_clock_info)
+{
+	uint16_t mclk_multiple = 256;
+
+	if (i2s_cfg == NULL) {
+		LOG_ERR("Input i2s_cfg is NULL");
+		return ESP_ERR_INVALID_ARG;
+	}
+
+	if (i2s_hal_clock_info == NULL) {
+		LOG_ERR("Input hal_clock_info is NULL");
+		return ESP_ERR_INVALID_ARG;
+	}
+
+	if (i2s_cfg->word_size == 24) {
+		mclk_multiple = 384;
+	}
+
+	if (i2s_cfg->options & I2S_OPT_FRAME_CLK_SLAVE ||
+	    i2s_cfg->options & I2S_OPT_BIT_CLK_SLAVE) {
+		i2s_hal_clock_info->bclk_div = 8;
+		i2s_hal_clock_info->bclk = i2s_cfg->frame_clk_freq * i2s_cfg->channels *
+			channel_length;
+		i2s_hal_clock_info->mclk = i2s_cfg->frame_clk_freq * i2s_hal_clock_info->bclk_div;
+	} else {
+		i2s_hal_clock_info->bclk = i2s_cfg->frame_clk_freq * i2s_cfg->channels *
+			channel_length;
+		i2s_hal_clock_info->mclk = i2s_cfg->frame_clk_freq * mclk_multiple;
+		i2s_hal_clock_info->bclk_div = i2s_hal_clock_info->mclk	/ i2s_hal_clock_info->bclk;
+	}
+
+	i2s_hal_clock_info->sclk = i2s_esp32_get_source_clk_freq(I2S_ESP32_CLK_SRC);
+	i2s_hal_clock_info->mclk_div = i2s_hal_clock_info->sclk / i2s_hal_clock_info->mclk;
+	if (i2s_hal_clock_info->mclk_div == 0) {
+		LOG_ERR("Sample rate is too large for the current clock source");
+		return ESP_ERR_INVALID_ARG;
+	}
+
+	return ESP_OK;
+}
+
+static void i2s_esp32_queue_drop(struct i2s_esp32_stream *stream)
+{
+	struct queue_item item;
+
+	while (k_msgq_get(&stream->queue, &item, K_NO_WAIT) == 0) {
+		k_mem_slab_free(stream->i2s_cfg.mem_slab, item.buffer);
+	}
+}
+
+static void i2s_esp32_rx_callback(const struct device *dma_dev, void *arg, uint32_t channel,
+				  int status);
+static void i2s_esp32_tx_callback(const struct device *dma_dev, void *arg, uint32_t channel,
+				  int status);
+
+static int i2s_esp32_restart_dma(const struct device *dev, enum i2s_dir i2s_dir)
+{
+	const struct i2s_esp32_cfg *dev_cfg = dev->config;
+	struct i2s_esp32_data *const dev_data = dev->data;
+	struct i2s_esp32_stream *stream;
+	void *src = NULL, *dst = NULL;
+	int err;
+
+	if (i2s_dir == I2S_DIR_RX) {
+		stream = &dev_data->rx;
+		dst = stream->mem_block;
+	} else if (i2s_dir == I2S_DIR_TX) {
+		stream = &dev_data->tx;
+		src = stream->mem_block;
+	} else {
+		LOG_ERR("Invalid DMA direction");
+		return -EINVAL;
+	}
+
+	err = dma_reload(stream->dma_dev, stream->dma_channel, (uint32_t)src, (uint32_t)dst,
+				stream->mem_block_len);
+	if (err < 0) {
+		LOG_ERR("Error reloading DMA channel[%d]: %d", (int)stream->dma_channel, err);
```

Comment:
@kartben Must all these checks be addressed? Sounds there are one's that does not improve code reading. What is current policy regarding it?

Reply:
this was introduced with https://github.com/zephyrproject-rtos/zephyr/pull/75089
afaict they are warnings/FYI only but please chime in in the PR if you have suggestions on maybe improving the wording so that there's less chances submitter/reviewers get confused and end up spending time stuff that's not strictly necessary or not helping with code readability.